Add comment on reblog in follow-plugin #758

maslenitsa93 opened the pull request:
https://github.com/GolosChain/golos/issues/758

Checklist:
1. Create 3 accounts and a comment to reblog:
```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
create_account cyberfounder test "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder test "30.000 GOLOS" true
post_comment  test test "" test hello world "{}" true

create_account cyberfounder testreb "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder testreb "30.000 GOLOS" true
create_account cyberfounder testreb2 "{}" "30.000 GOLOS" true
transfer_to_vesting cyberfounder testreb2 "30.000 GOLOS" true
```

2. Follow on future rebloggers (need for feed_objects):
```
begin_builder_transaction
add_operation_to_builder_transaction 0 ["custom_json", {"required_posting_auths":["test"],"id": "follow", "json":"[\"follow\", {\"follower\":\"test\",\"following\":\"testreb\",\"what\":[\"blog\"]}]"}]
sign_builder_transaction 0 true
begin_builder_transaction
add_operation_to_builder_transaction 1 ["custom_json", {"required_posting_auths":["test"],"id": "follow", "json":"[\"follow\", {\"follower\":\"test\",\"following\":\"testreb2\",\"what\":[\"blog\"]}]"}]
sign_builder_transaction 1 true
```

3. Reblog
```
begin_builder_transaction
add_operation_to_builder_transaction 2 ["custom_json", {"required_posting_auths":["testreb"],"id": "follow", "json":"[\"reblog\", {\"account\":\"testreb\",\"author\":\"test\",\"permlink\":\"test\",\"title\":\"hello\",\"body\":\"world\",\"json_metadata\":\"{}\"}]"}]
sign_builder_transaction 2 true
begin_builder_transaction
add_operation_to_builder_transaction 3 ["custom_json", {"required_posting_auths":["testreb2"],"id": "follow", "json":"[\"reblog\", {\"account\":\"testreb2\",\"author\":\"test\",\"permlink\":\"test\",\"title\":\"hello2\",\"body\":\"world2\"}]"}]
sign_builder_transaction 3 true
```

4. Check blog_index.
```
</script><script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
var logger = function(err, result) { console.log(err, result); };
golos.api.getBlog("test", 0, 300, function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
golos.api.getBlog("testreb", 0, 300, function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
golos.api.getBlog("testreb2", 0, 300, function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result: 3 blogs - by test, testreb, testreb2. 
testreb's and testreb2's blogs have `title` and `body` fields. Second has empty `json_metadata` field.
Also, try `getBlogEntries` - result should be same.

5. Check feed_index:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
var logger = function(err, result) { console.log(err, result); };
golos.api.getFeed("test", 0, 300, function (err, result) {
document.body.innerHTML = '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
```
Result: feed_object with testreb and testreb2 in `reblog_by` field, and with `reblog_entries` field with 2 reblog entries.
Also, try `getFeedEntries` - result should be same.

6. Check get_discussions_by_blog:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
var logger = function(err, result) { console.log(err, result); };
golos.api.getDiscussionsByBlog({select_authors: ["test"], limit: 100}, function (err, result) {
document.body.innerHTML = '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
```
Result: discussion with reblog entry.

7. Check get_discussions_by_feed:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
var logger = function(err, result) { console.log(err, result); };
golos.api.getDiscussionsByFeed({select_authors: ["test"], limit: 100}, function (err, result) {
document.body.innerHTML = '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
```
Result: discussion with reblog entries.

afalaleev commented on plugins/follow/include/golos/plugins/follow/follow_operations.hpp:
```diff
@@ -24,6 +24,9 @@ namespace golos {
                 protocol::account_name_type account;
                 protocol::account_name_type author;
                 std::string permlink;
+                string title;
```
Add fields to FC_REFLECT.

afalaleev commented on plugins/tags/plugin.cpp:
```diff
@@ -560,6 +560,9 @@ namespace golos { namespace plugins { namespace tags {
                     d.first_reblogged_on = b.reblogged_on;
+                    d.title = to_string(b.title);
+                    d.body = to_string(b.body);
+                    d.json_metadata = to_string(b.json_metadata);
```
Add other fields for blog. These fields is already is filled by comment.

maslenitsa93 commented:
Added validation for fields `title`, `body`, `json_metadata`. `body` is required if one of these 3 fields is set. All another fields are optional.

Added filling of fields in `get_feed` and `get_feed_entries`.

afalaleev commented on libraries/api/include/golos/api/reblog_entry.hpp:
```diff
@@ -0,0 +1,22 @@
+        std::string body;
+        std::string json_metadata;
+
+        reblog_entry() {
```
`= default;`

afalaleev commented on libraries/api/include/golos/api/reblog_entry.hpp:
```diff
@@ -0,0 +1,22 @@
+    struct reblog_entry {
+        std::string title;
+        std::string body;
+        std::string json_metadata;
```
Add ab author of reblog.

afalaleev commented on libraries/api/include/golos/api/reblog_entry.hpp:
```diff
@@ -0,0 +1,22 @@
+        reblog_entry() {
+        }
+
+        reblog_entry(const std::string& title_, const std::string& body_, const std::string& json_metadata_)
```
May be initialize from blog_object? It will require less copy-paste at calling places.

maslenitsa93 replied:
@afalaleev blog_object is in plugin, reblog_entry is in api-library (couldn't be moved to plugin because used in discussion class).
Also, get_feed and get_feed_entries are initially big copy-paste, and it isn't single case.

afalaleev commented on plugins/follow/include/golos/plugins/follow/follow_api_object.hpp:
```diff
@@ -31,13 +35,19 @@ namespace golos {
                 uint32_t entry_id = 0;
+                std::string title;
+                std::string body;
+                std::string json_metadata;
```
Rename to reblog_XXX.
It may mislead with title, body and json from comment.
